Firestore/core/src/util/random_access_queue.h

```diff
@@ -0,0 +1,265 @@
+ *
+ * Another way to think of this class is an `unordered_set` that preserves
+ * insertion order and provides efficient access to the "front" element and
+ * removal of arbitrary elements.
+ *
+ * This class is implemented by maintaining two data structures: a queue that
+ * keeps track of insertion order and a set that provides efficient lookup.
+ * As a result, this class consumes roughly twice as much memory as a plain old
+ * queue and each mutating operation is roughly twice as slow as a plain old
+ * queue since each mutation must be performed on two data structures. However,
+ * these costs come at the benefit of being able to test for queue membership in
+ * constant time and remove queue elements in amortized constant time.
+ *
+ * The template argument `T` is the type of the elements to store in the queue.
+ * The remaining variadic template arguments are passed as the template
+ * arguments to `unordered_map`, allowing custom hashing and comparison
+ * functions to be specified.
+ */
+template <typename T, typename... UnorderedMapArgs>
+class RandomAccessQueue {
+ public:
+  RandomAccessQueue() = default;
+
+  RandomAccessQueue(const RandomAccessQueue& other) {
+    PushBackAll(other);
+  }
+
+  RandomAccessQueue(RandomAccessQueue&& other) = default;
+
+  RandomAccessQueue& operator=(const RandomAccessQueue& other) {
+    queue_.clear();
+    queue_entries_by_element_.clear();
+    PushBackAll(other);
+    return *this;
+  }
+
+  RandomAccessQueue& operator=(RandomAccessQueue&& other) = default;
+
+  /**
+   * Adds an element to the back of this queue, if it is not already present.
+   *
+   * Returns `true` if the given element was *not* already present in this
+   * queue, and therefore was added to the back, or `false` if the given element
+   * *was* already present in this queue, and therefore no changes were made to
+   * this object.
+   *
+   * This method has average constant-time complexity.
+   */
+  bool push_back(const T& element) {
+    // Put the element into the map if it is not already present. The `nullptr`
+    // will be replaced by a pointer to the object added to the queue when it is
+    // added below.
+    auto map_emplace_result =
+        queue_entries_by_element_.emplace(element, nullptr);
+
+    // Return `false` and do nothing if the given element was already present.
+    if (!map_emplace_result.second) {
+      return false;
+    }
+
+    // Add the element to the queue and update the pointer in the map.
+    queue_.emplace_back(element);
+    map_emplace_result.first->second = &queue_.back();
+
+    // Return `true` to indicate that the element was added to the queue.
+    return true;
+  }
+
+  /**
+   * Returns the element at the front of the queue.
+   *
+   * The behavior of this method is undefined if the queue is empty.
+   *
+   * This method has constant-time complexity.
+   */
+  const T& front() const {
+    return queue_.front().element();
```

Comment:
Should we assert here that `removed` is set to false?

Reply:
Done.